Add data parallel support to ModelTester, and some data parallel tests to CI

hshahTT opened the pull request:
### Ticket
[590](https://github.com/tenstorrent/tt-torch/issues/590)

### Problem description
Need to add support to run our test models in full execution mode using data parallel execution.

### What's changed
At a high level this PR modifies the `ModelTester` class to use the async mode functions added in [PR 696](https://github.com/tenstorrent/tt-torch/pull/696) to run the same input on multiple predefined devices in parallel.

Specifically:

- **tests/utils.py**:
    - Added two new member parameters to the `ModelTester` class:
        - `devices`: An optional list of devices to run the inputs on. If no device list is provided, the tester object will acquire all devices on the board and use them.
        - `data_parallel_mode`: Boolean, if True, will run the input on all of the provided devices.
    - Added a function to compile the given framework model for all provided devices (`compile_models_for_data_parallel()`)
    - Added a function (`_test_model_eval_data_parallel()`) to run the inputs asynchronously on all devices, and compare each output to the golden.
        - As a result of this, the `test_model()` function will return a list of outputs (one for each provided device) when running the test in data parallel mode.

- **tests/models/**:
    - Parameterized the MNIST, ResNet-50, and ResNet-18 model tests with a new `data_parallel_mode` parameter.

- **.github/workflows/run-full-model-execution-tests.yml**:
    - Added the data parallel versions of the MNIST, ResNet-50, and ResNet-18 model tests to the on-pr and on-push workflows. 

- All other file changes are no-ops and were made to ensure that they reflected the new `data_parallel_mode` parameter in the model test files.

### Notes
- Op by Op tests and runtime intermediate verification are not supported with data parallel mode.
- Once this PR is approved and submitted, I'll add more PRs to add other models to CI.

### Checklist
- [x] New/Existing tests provide coverage for changes

hshahTT committed:
[in progress] adding async support to ModelTester

hshahTT committed:
Finish data parallel testing implementation and try mnist model

hshahTT committed:
Add mnist to CI

hshahTT committed:
Add resnet50 and resnet18 to CI

hshahTT commented:
@kmabeeTT So far I've only added the MNIST, Resnet-50, and Resnet-18 models to the On-PR/On-Push CI since they don't take long to run (about ~30s - 1 min each). Could you let me know what models are worth adding to the nightlies and/or weeklies CIs? (i.e. which of the models in those workflows are the most important)

Also, how many data parallel tests should we aim to have in the CIs in total? I don't wanna make the workflows take too long.

hshahTT committed:
Fix failing distilbert test

ddilbazTT commented on tests/models/mnist/test_mnist.py:
```diff
@@ -74,17 +78,25 @@ def test_mnist_train(record_property, mode, op_by_op):
     cc.enable_consteval = True
     cc.consteval_parameters = True
     if op_by_op:
+        if data_parallel_mode:
```
Is data parallel ever going to be supported during op-by-op? If never, maybe there should be an assert somewhere, probably `backend.py`

hshahTT replied:
I had previously added an assert in Executor's constructor: https://github.com/tenstorrent/tt-torch/blob/813ea5d7a0ca5d4975dedd1c996570f1d979a1e9/tt_torch/dynamo/executor.py#L188

ddilbazTT commented on tests/models/mnist/test_mnist.py:
```diff
@@ -74,17 +78,25 @@ def test_mnist_train(record_property, mode, op_by_op):
             cc.op_by_op_backend = OpByOpBackend.STABLEHLO
-
+    devices = None
+    if data_parallel_mode:
```
Can't we do these in the ModelTester?

hshahTT replied:
We could. I wanted to give more flexibility to the user on selecting which devices to use, rather than just defaulting to all devices. For example, when we expand this to LLMBox tests, then we can write tests that run Model A on 4 devices and Model B on the remaining 4 devices.

hshahTT committed:
No longer require user to provide list of devices for data parallel mode. If no list is provided, all devices on the board will be acquired and managed by the ModelTester object.

hshahTT commented:
@ddilbazTT @jameszianxuTT I modified the `ModelTester` class so that the device list is now optional. If no device list is provided, the tester object will just acquire all devices on the board and use them for data parallel. I also updated the PR description to reflect this.

hshahTT commented:
Single device vs Data Parallel Benchmarking results (10 testing runs each):

**MNIST Test:**
<img width="1210" alt="image" src="https://github.com/user-attachments/assets/71aeba33-73bc-48f8-a2eb-5e6aadbbfb0e" />

**Resnet 18 Test:**
<img width="1207" alt="image" src="https://github.com/user-attachments/assets/cce85040-2928-4dda-9819-04fbf5929213" />

**Resnet 50 Test:**
<img width="1216" alt="image" src="https://github.com/user-attachments/assets/7c219e79-eaf8-4649-9984-342f6c0e35dc" />

hshahTT committed:
Release devices in test_pipeline_parallel() unit test

AleksKnezevic commented on .github/workflows/run-full-model-execution-tests.yml:
```diff
@@ -27,7 +27,7 @@ env:
 jobs:
   tests:
-    timeout-minutes: 60 # current onPR time runs ~60m
+    timeout-minutes: 63 # current onPR time runs ~62-63m
```
you should have some leeway here, I would go with 90

hshahTT committed:
Increase timeout

ddilbazTT reviewed (commented):
I have a few remarks to add. Thanks for addressing the PR comments

hshahTT committed:
remove data parallel check in tests

jameszianxuTT commented:
Crap, I see adding this extra parameterization makes the assumption we had before that cli option --op_by_op_* would run the single op by op test defined for the model, but now it splays out into +/- data parallel

We're gonna end up with a combinatorial explosion of different parallelization / multichip modes later - I am fine with just getting this PR in but we'll need to think of a proper testing strategy for multichip tests #780 . 

I'll push a quick fix to unblock this pr as we have some other stuff coming down the pipeline @hshahTT

jameszianxuTT committed:
Filter only single device obo shlo parameterization in test yaml

hshahTT commented on tests/models/mnist/test_mnist.py:
```diff
@@ -10,6 +10,7 @@
 import torch.nn.functional as F
 from tests.utils import ModelTester
 from tt_torch.tools.utils import CompilerConfig, CompileDepth, OpByOpBackend
+from tt_torch.tools.device_manager import DeviceManager
```
Good catch! Removed it

hshahTT committed:
Add a common function for verifying full execution outputs, enable support for --op_by_op_torch and --op_by_op_stablehlo flags with data parallel parameterization by adding py.skip statements

vmilosevic commented:
<table><tr><th><th>Tests</th><th>Passed ✅</th><th>Skipped ⚠️</th><th>Failed</th></tr><tr><td>TT-Torch Tests</td><td>763 ran</td><td>748 passed</td><td>15 skipped</td><td>0 failed</td></tr></table>

<table><tr><th>Test</th><th>Result</th></tr><tr><td colspan="2">No test annotations available</td></tr></table>

<!-- Summary comment for ["TT-Torch Tests"] by mikepenz/action-junit-report -->
